crates/ruff_cli/src/printer.rs

```diff
@@ -439,3 +434,71 @@ fn print_fix_summary(writer: &mut dyn Write, fixed: &FxHashMap<String, FixTable>
     }
     Ok(())
 }
+
+/// Contains the number of [`Applicability::Automatic`] and [`Applicability::Suggested`] fixes
+struct FixableStatistics<'a> {
+    automatic: u32,
+    suggested: u32,
+    apply_suggested: &'a SuggestedFixes,
+}
+
+impl<'a> FixableStatistics<'a> {
+    fn new(diagnostics: &Diagnostics, apply_suggested: &'a SuggestedFixes) -> Self {
+        let mut automatic = 0;
+        let mut suggested = 0;
+
+        for message in &diagnostics.messages {
+            if let Some(fix) = &message.fix {
+                if fix.applicability() == Applicability::Suggested {
+                    suggested += 1;
+                } else if fix.applicability() == Applicability::Automatic {
+                    automatic += 1;
+                }
+            }
+        }
+
+        Self {
+            automatic,
+            suggested,
+            apply_suggested,
+        }
+    }
+
+    fn fixes_are_applicable(&self) -> bool {
+        match self.apply_suggested {
+            SuggestedFixes::Apply => self.automatic > 0 || self.suggested > 0,
+            SuggestedFixes::Disable => self.automatic > 0,
+        }
+    }
+
+    /// Returns [`true`] if there aren't any fixes to be displayed
+    fn is_empty(&self) -> bool {
+        self.automatic == 0 && self.suggested == 0
+    }
+
+    /// Build the displayed fix status message depending on the types of the remaining fixes.
+    fn violation_string(&self) -> String {
+        let automatic_prefix = format!("[{}]", Applicability::Automatic.symbol().cyan());
+        let suggested_prefix = format!("[{}]", Applicability::Suggested.symbol().cyan());
+
+        if self.automatic > 0 && self.suggested > 0 {
+            format!(
+                "{automatic_prefix} {} fixable with the --fix option.\n\
+                {suggested_prefix} {} potentially fixable with the --fix-suggested option.",
+                self.automatic, self.suggested
+            )
+        } else if self.automatic > 0 {
+            format!(
+                "{automatic_prefix} {} fixable with the --fix option.",
+                self.automatic,
+            )
+        } else if self.suggested > 0 {
+            format!(
+                "{suggested_prefix} {} potentially fixable with the --fix-suggested option.",
+                self.suggested
+            )
+        } else {
+            String::new()
```

Comment:
This case combined with `is_empty` feels slightly off to me, since if someone is calling `violation_string` on an empty set, it's probably a mistake? I could see a few alternatives...

1. `fn new` could be `fn try_from` or something like that, and it could return `Option<Self>` (`None` when there are no fixes), so that it's impossible to represent an empty set of fixes.
2. This method could return an `Option` (and `is_empty` could be removed).

(1) feels the most natural to me. What do you think?